### PR TITLE
Gradient ascent

### PR DESCRIPTION
**Pull request by Lucaszw**
Do Not Merge! 
Gradient Ascent for Caffe (Torch support not completed yet)
Depends on #937

Optimization Technique From Deep Vis Toolbox
```
@inproceedings{yosinski-2015-ICML-DL-understanding-neural-networks,
Author = {Jason Yosinski and Jeff Clune and Anh Nguyen and Thomas Fuchs and Hod Lipson},
Booktitle = {Deep Learning Workshop, International Conference on Machine Learning (ICML)},
Title = {Understanding Neural Networks Through Deep Visualization},
Year = {2015}}
```

### CODE REVIEW

**Comment by lukeyeager** [comment]
Can you add some screenshots so people can get a sense of how cool this will be?

**Comment by Lucaszw** [comment]
cifar10

**Comment by Lucaszw** [comment]
googlenet_pretrained

**Comment by TimZaman** [comment]
OMG! That's a really cool idea, and it looks really really sexy. I bet this will be very useful getting some deeper insight ;).

**Comment by Lucaszw** [comment]
Thanks @TimZaman ! And yeah drag-n-droppable layers would be epic!! I will check out jsplumb tonight :+1:

**Comment by gheinrich** [comment]
Did I miss a dependency or is `tools/gradient_ascent/caffe_misc.py` missing?

**Comment by gheinrich** [comment]
It would be nice to publish a walk-through to show how to use this.

**Comment by Lucaszw** [comment]
@gheinrich . My bad, that should no longer a dependency! I removed the import statement in my latest commit :)

**Comment by Lucaszw** [comment]
@gheinrich A walk-through sounds like a great idea!

**Comment by gheinrich** [comment]
I get this error:
```
2016-08-22 18:21:42 [ERROR] TypeError: "axis" are not valid keyword arguments.
2016-08-22 18:21:42 [20160822-182134-1ab3] [WARNING] Get Max Activations From Pretrained Model unrecognized output: Traceback (most recent call last):
2016-08-22 18:21:42 [20160822-182134-1ab3] [WARNING] Get Max Activations From Pretrained Model unrecognized output: File "/home/greg/ws/digits/tools/gradient_ascent_caffe.py", line 267, in <module>
2016-08-22 18:21:42 [20160822-182134-1ab3] [WARNING] Get Max Activations From Pretrained Model unrecognized output: args['gpu']
2016-08-22 18:21:42 [20160822-182134-1ab3] [WARNING] Get Max Activations From Pretrained Model unrecognized output: File "/home/greg/ws/digits/tools/gradient_ascent_caffe.py", line 142, in infer
2016-08-22 18:21:42 [20160822-182134-1ab3] [WARNING] Get Max Activations From Pretrained Model unrecognized output: y = np.sum(np.square(np.gradient(np.mean(im, axis=2),axis=1)), axis=1)
```
I think the `axis` keyword in `numpy.gradient()` was added in Numpy 1.11.0. I have 1.8.2, which is "legal" as per the requirements.txt file. Shall we update the requirements or can you do without this keyword?

**Comment by Lucaszw** [comment]
@gheinrich Shoot I didn't realize that! Unfortunately I do need the keyword, but I can look for an alternative such that it works for 1.8 and 1.11

File: digits/framework_helpers/caffe_helpers.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+    # TODO - Implement crop_size
+    shape.dim.append(feature_dims[0])
+    shape.dim.append(feature_dims[1])
+
+    # Layers
+    deploy_network.MergeFrom(deploy_layers)
+
+    # Write to file
+    with open(path+"/"+deploy_file, 'w') as outfile:
```
Comment (TimZaman) [inline digits/framework_helpers/caffe_helpers.py]: os.join

File: digits/model/views.py [hunk 2]
```diff
@@ -168,6 +178,62 @@ def visualize_lr():
...
+
+    task = job.train_task()
+    snapshot_filename = None
+    snapshot_filename = task.get_snapshot(epoch)
+
+    job = PretrainedModelJob(
+        snapshot_filename,
+        job.dir() + "/" + task.model_file ,
+        task.dataset.dir() + "/" + info["labels file"],
+        task.dataset.dir() + "/" + info["mean file"],
```
Comment (TimZaman) [inline digits/model/views.py]: os.join x3

File: digits/pretrained_model/tasks/upload_pretrained.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    def get_labels(self):
+        labels = []
+        if self.labels_path is not None:
+            with open(self.job_dir+"/labels.txt") as f:
+                labels = f.readlines()
+        return labels
+
+    def move_file(self,input, output,env):
+        args  = ["cp", input, self.job_dir+"/"+output]
```
Comment (TimZaman) [inline digits/pretrained_model/tasks/upload_pretrained.py]: os.join

File: digits/pretrained_model/tasks/upload_pretrained.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                            self.gpu = int(resource.identifier)
+                            reserved_resources[gpu_key] = [(resource.identifier, 1)]
+                            break
+                return reserved_resources
+        return None
+
+    def get_labels(self):
+        labels = []
+        if self.labels_path is not None:
+            with open(self.job_dir+"/labels.txt") as f:
```
Comment (TimZaman) [inline digits/pretrained_model/tasks/upload_pretrained.py]: os.join

**Comment by TimZaman** [comment]
Yo Lucas. Like the other PR, there are round 40 instances where concatenated paths -> os.join.

**Comment by Lucaszw** [comment]
@TimZaman  I haven't updated this PR in a while (I've been re-basing locally so they should be gone next time I update)

**Comment by lukeyeager** [comment]
Closing for now (there are merge conflicts anyway). I'd like to revisit this.

I've backed up this work at https://github.com/lukeyeager/DIGITS/tree/backup/lucas/gradientAscent.

**Pull request closed without merging**